src/LinuxTracing/LibunwindstackUnwinder.cpp

```diff
@@ -116,7 +125,99 @@ LibunwindstackResult LibunwindstackUnwinderImpl::Unwind(
 
   return LibunwindstackResult{unwinder.ConsumeFrames(), unwinder.LastErrorCode()};
 }
+// This functions detects if a frame pointer register was set in the given program counter using
+// the given Dwarf section.
+// It does so by verifying if "Canonical Frame Address" gets computed immediately from $rbp (with
+// offset 16 to skip the old frame pointer and the return address).
+// The function returns nullopt if the required Dwarf information is not available.
+static std::optional<bool> HasFramePointerSetFromDwarfSection(
+    uint64_t rel_pc, unwindstack::DwarfSection* dwarf_section,
+    std::map<uint64_t, unwindstack::DwarfLocations>* loc_regs_cache) {
+  if (dwarf_section == nullptr) {
+    return false;
+  }
+
+  auto cache_it = loc_regs_cache->upper_bound(rel_pc);
+  if (cache_it == loc_regs_cache->end() || rel_pc < cache_it->second.pc_start) {
+    const unwindstack::DwarfFde* fde = dwarf_section->GetFdeFromPc(rel_pc);
+    if (fde == nullptr) {
+      return std::nullopt;
+    }
+    unwindstack::DwarfLocations loc_regs;
+    if (!dwarf_section->GetCfaLocationInfo(rel_pc, fde, &loc_regs, unwindstack::ARCH_X86_64)) {
+      return std::nullopt;
+    }
+    cache_it = loc_regs_cache->emplace(loc_regs.pc_end, std::move(loc_regs)).first;
+  }
+
+  unwindstack::DwarfLocations* loc_regs = &cache_it->second;
+  auto cfa_entry = loc_regs->find(unwindstack::CFA_REG);
+  if (cfa_entry == loc_regs->end()) {
+    return false;
+  }
+  const unwindstack::DwarfLocation* loc = &cfa_entry->second;
+  ORBIT_CHECK(loc != nullptr);
+  if (loc->type == unwindstack::DWARF_LOCATION_REGISTER) {
+    // From the Dwarf standard:
+    //  "Typically, the CFA is defined to be the value of the stack pointer at the call site in the
+    //  previous frame (which may be different from its value on entry to the current frame)"
+    // So for the frame pointer case, the "value of the stack pointer at the call site" is:
+    // $rbp + 8 (for the prev. frame pointer) + 8 (for the return address)
+    if (loc->values[0] == unwindstack::X86_64_REG_RBP && loc->values[1] == 16) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+std::optional<bool> LibunwindstackUnwinderImpl::HasFramePointerSet(uint64_t instruction_pointer,
+                                                                   unwindstack::Maps* maps) {
+  unwindstack::MapInfo* map_info = maps->Find(instruction_pointer);
+  if (map_info == nullptr) {
+    return std::nullopt;
+  }
+
+  // We don't have enough process memory collected to read the object file from memory (if memory
+  // mapped), so we don't even try and directly read from file.
+  unwindstack::Object* object = map_info->GetObject(nullptr, unwindstack::ARCH_X86_64);
+  if (object == nullptr) {
+    return std::nullopt;
+  }
+
+  auto* elf = dynamic_cast<unwindstack::Elf*>(object);
+  if (elf == nullptr) {
+    // TODO(b/228599622): Handle the PeCoff case.
```

Comment:
Many of the PE/COFF files that are relevant for use (Wine) have a .debug_frame section, so it may be possible to combine the cases to some degree.